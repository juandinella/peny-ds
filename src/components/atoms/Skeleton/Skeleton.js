import React from 'react'
import PropTypes from 'prop-types'
import styles from './Skeleton.styles'

/**
 * Display a skeleton whilst another component is loading.
 */

const Skeleton = ({ id, width, height, rounded, squared, children, show }) => {
  return show ? (
    <div
      css={styles}
      id={id}
      data-testid={id}
      rounded={rounded}
      squared={squared}
      style={{ width: width, height: height }}
      className={width ? 'hasWidth' : null}
    >
      {children}
    </div>
  ) : (
    children
  )
}

Skeleton.displayName = 'Skeleton'

Skeleton.propTypes = {
  /** Skeleton's id */
  id: PropTypes.string,
  /**
   * Skeleton's Width.
   *
   * If there's no width, it will adjust automatically to the children
   */
  width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  /**
   * Skeleton's Height
   *
   * Min Height 20px;
   */
  height: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  /** Rounded Skeleton */
  rounded: PropTypes.bool,
  /** Squared Skeleton */
  squared: PropTypes.bool,
  /** Content to be included in the skeleton */
  children: PropTypes.node,
  /** If true skeleton should be shown, otherwise only children should be shown */
  show: PropTypes.bool,
}

Skeleton.defaultProps = {}

export default Skeleton
