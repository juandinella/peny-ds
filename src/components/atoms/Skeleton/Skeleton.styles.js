import { css } from 'styled-components'

export default css`
  position: relative;
  min-height: 20px;
  display: block;

  &:before {
    content: '';
    position: absolute;
    top: -1px;
    left: -1px;
    bottom: 0;
    z-index: 5;
    height: calc(100% + 2px);
    width: calc(100% + 2px);
    border-radius: var(--radius-sm);
    background-image: linear-gradient(
      270deg,
      var(--color-neutral-00),
      var(--color-neutral-200),
      var(--color-neutral-200),
      var(--color-neutral-00)
    );
    background-size: 400% 100%;
    animation: skeletonShine 8s ease-in-out infinite;
  }

  @keyframes skeletonShine {
    0% {
      background-position: 200% 0;
    }
    to {
      background-position: -200% 0;
    }
  }

  ${(props) =>
    props.squared &&
    css`
      border-radius: 0;
    `}

  ${(props) =>
    props.rounded &&
    css`
      border-radius: 100%;
    `}

  &.hasWidth {
    min-height: 24px;
    border-radius: var(--radius-sm);
    background-image: linear-gradient(
      270deg,
      var(--color-neutral-00),
      var(--color-neutral-200),
      var(--color-neutral-200),
      var(--color-neutral-00)
    );
    background-size: 400% 100%;
    animation: skeletonShine 8s ease-in-out infinite;

    &:before {
      content: none;
    }

    ${(props) =>
      props.squared &&
      css`
        border-radius: 0;
      `}

    ${(props) =>
      props.rounded &&
      css`
        border-radius: 100%;
      `}
  }
`
