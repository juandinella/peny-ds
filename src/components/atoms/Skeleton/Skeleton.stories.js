import Skeleton from './Skeleton'

import { getTemplate } from '../../../helpers/storybook'

const Template = getTemplate(Skeleton)

export default {
  title: 'Atoms/Skeleton',
  component: Skeleton,
  args: {
    width: 124,
    show: true,
  },
  argTypes: {
    width: { control: 'number' },
    height: { control: 'number' },
  },
}

export const Default = Template.bind({})
