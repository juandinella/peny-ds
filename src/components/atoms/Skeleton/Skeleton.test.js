import React from 'react'

import { cleanup, render } from '@testing-library/react'

import Skeleton from './Skeleton'

import initStoryshots from '@storybook/addon-storyshots'

initStoryshots({
  storyKindRegex: /^Atoms\/Skeleton$/,
})

const componentProps = {
  id: 'skeleton-id',
  show: true,
  children: <div data-testid="children">Children</div>,
}

afterEach(cleanup)

describe('<Skeleton />', () => {
  test('should render component correctly', () => {
    const { queryByTestId } = render(<Skeleton {...componentProps} />)

    const skeleton = queryByTestId(componentProps.id)

    expect(skeleton).toBeTruthy()
  })

  test('should render only children', () => {
    const { queryByTestId } = render(
      <Skeleton {...componentProps} show={false} />
    )

    const skeleton = queryByTestId(componentProps.id)
    const children = queryByTestId('children')

    expect(skeleton).toBeFalsy()
    expect(children).toBeTruthy()
  })
})
