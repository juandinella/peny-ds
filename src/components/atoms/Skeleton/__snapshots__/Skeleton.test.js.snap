// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Atoms/Skeleton Default 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <div
    className="sc-kjnynt-0 eneJQj hasWidth"
    style={
      Object {
        "height": undefined,
        "width": 124,
      }
    }
  />
</div>
`;
